File: cpp/libs/gubg/Cartouche.cpp
#include "gubg/Cartouche.hpp"
#include "gubg/parse/Line.hpp"
using namespace std;

#define GUBG_MODULE_ "Cartouche"
#include "gubg/log/begin.hpp"
namespace  { 
    const string marker("!@#");
} 

namespace gubg { 

    Cartouche::Cartouches Cartouche::parse(const string &content)
    {
        S();
        Cartouches res;
        Cartouche *current = 0;
        string pre;
        for (auto &l: line::split(content))
        {
            SS(l.txt);
            if (!current)
            {
                L("No cartouche detected yet");
                const auto ix = l.txt.find(marker);
                if (ix == string::npos)
                    //Still no cartouche detected
                    continue;
                L("We detected a cartouche, add one and initialize it");
                res.push_back(Cartouche());
                current = &res.back();
                current->tag = l.txt.substr(ix+marker.size());
                pre = l.txt.substr(0, ix);
            }
            else
            {
                L("We are detecting a cartouche");
                if (pre != l.txt.substr(0, pre.size()))
                {
                    LL("The pre does not match anymore, the previous cartouche is closed now");
                    current = 0;
                    continue;
                }
                L("This line is part of the cartouche, add it, together with its endline marker");
                current->content.append(l.txt.substr(pre.size()));
                current->content.append(to_s(l.end));
            }
        }
        return res;
    }

}
#include "gubg/log/end.hpp"

File: rust/gubg/cartouche.rs
use std::io::stdio::println;

pub struct Cartouche
{
    pub tag: ~str,
    pub lines: ~[~str],
}

impl Cartouche
{
    pub fn new() -> Cartouche
    {
        Cartouche{tag: ~"", lines: ~[]}
    }
    pub fn parse(content: &str) -> ~[~Cartouche]
    {
        println!("Content: {}", content);
        let mut res = ~[];
        res
    }
    pub fn print(&self)
    {
        println(self.tag);
        for line in self.lines.iter()
        {
            println!("\t{}", *line);
        }
    }
}

pub struct Cartouches<'a>
{
	content: ~str,
	pub cartouches: ~[&'a str],
}

impl<'a> Cartouches<'a>
{
	pub fn from_str(content: ~str) -> Cartouches
	{
		let mut res = Cartouches{content: content, cartouches: ~[]};
		let mut ix = 0;
		for line in res.content.lines_any()
		{
			ix += 1;
			println!("{}: {}", ix, line);
			res.cartouches.push(line);
		}
		Cartouches{content: ~"bla", cartouches: ~[]}
	}
}
